### PR TITLE
Move PingType metadata to the Rust side

### PR DESCRIPTION
**Pull request by mdboom**
I'm just posting this here for discussion -- either here or in our next meeting if that's more efficient.  FFI and Kotlin side not implemented yet.

I first considered putting the ping registry on the Glean object, since it sort of logically makes sense there.  However, libraries need to register pings at any time during startup, including before `Glean::initialize` is called.  Since we've tied `new` and `initialize` together, we couldn't add pings to the ping registry before the Glean singleton is initialized, since before that there is no Rust-side Glean object.  An alternative to make that work would be to split `new` and `initialize` apart again, which isn't great.

So I came up with this solution which uses an actual global object and state on the Rust side.  It's not bad, but it does mean for testing that there is state carried over between tests.

Tl;dr: I thought this could use more feedback / brainstorming before filling out the rest of the pieces.

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

**Comment by mdboom** [comment]
@Dexterp37 , @badboy : Totally new approach as discussed the other day.  Ping types exist on the Glean object, but we queue them on the Kotlin side until `Glean.initialized` is called.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 1]
```diff
@@ -368,6 +380,30 @@ open class GleanInternalAPI internal constructor () {
         LibGleanFFI.INSTANCE.glean_destroy_glean(handle, e)
         handle = 0L
     }
+
+    /**
+     * Register a [PingType] in the registry associated with this [Glean] object.
+     */
+    internal fun registerPingType(pingType: PingType) {
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Does this need to be `@Synchronized` together with `initialize`?

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 2]
```diff
@@ -368,6 +380,30 @@ open class GleanInternalAPI internal constructor () {
...
+        if (handle == 0L) {
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: I'd use `isInitialized()` here.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 3]
```diff
@@ -368,6 +380,30 @@ open class GleanInternalAPI internal constructor () {
...
+            )
+        }
+    }
+
+    /**
+     * Returns true if a ping by this name is in the ping registry.
+     *
+     * For testing only.
+     */
+    // @VisibleForTesting(otherwise = VisibleForTesting.NONE)
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: nit: let's remove the commented line if not needed.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 4]
```diff
@@ -368,6 +380,30 @@ open class GleanInternalAPI internal constructor () {
...
+                handle,
+                pingType.handle
+            )
+        }
+    }
+
+    /**
+     * Returns true if a ping by this name is in the ping registry.
+     *
+     * For testing only.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: nit: "For internal testing only"

File: glean-core/src/lib.rs [hunk 5]
```diff
@@ -154,6 +157,41 @@ impl Glean {
...
         }
     }
+
+    /// Send a ping by name.
+    ///
+    /// The ping content is assembled as soon as possible, but upload is not
+    /// guaranteed to happen immediately, as that depends on the upload
+    /// policies.
+    ///
+    /// If the ping currently contains no content, it will not be sent.
```
Comment (Dexterp37) [inline glean-core/src/lib.rs]: nit: not be sent or not be assembled? :)

File: glean-core/src/lib.rs [hunk 6]
```diff
@@ -154,6 +157,41 @@ impl Glean {
...
+    ///
+    /// Returns true if a ping was sent, false otherwise.
```
Comment (Dexterp37) [inline glean-core/src/lib.rs]: nit: sent or assembled?

File: glean-core/src/lib.rs [hunk 7]
```diff
@@ -120,26 +124,25 @@ impl Glean {
...
 
-    /// Send a ping by name.
+    /// Send a ping.
     ///
     /// The ping content is assembled as soon as possible, but upload is not
     /// guaranteed to happen immediately, as that depends on the upload
     /// policies.
     ///
-    /// TODO: (Verify this is correct):
     /// If the ping currently contains no content, it will not be sent.
```
Comment (Dexterp37) [inline glean-core/src/lib.rs]: Since we're here, can we fix the language here? I think we need to use 'assembled' in most places in which we used 'sent'

File: glean-core/src/metrics/ping.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::error::Result;
+use crate::Glean;
+
+#[derive(Debug, Clone)]
+pub struct PingType {
```
Comment (Dexterp37) [inline glean-core/src/metrics/ping.rs]: Can we add documentation to this explaining what this is for?

File: glean-core/tests/ping.rs [hunk 9]
```diff
@@ -10,7 +10,10 @@ use glean_core::CommonMetricData;
 
 #[test]
 fn write_ping_to_disk() {
```
Comment (Dexterp37) [inline glean-core/tests/ping.rs]: Can we file a bug about porting over the tests from `PingMakerTest.kt` (from [glean-ac](https://github.com/mozilla-mobile/android-components/blob/master/components/service/glean/src/test/java/mozilla/components/service/glean/ping/PingMakerTest.kt)) to Rust? They cover most of the implementation here. Doesn't need to block this PR, it can be done as a follow-up.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 10]
```diff
@@ -77,6 +80,12 @@ open class GleanInternalAPI internal constructor () {
         this.gleanDataDir = File(applicationContext.applicationInfo.dataDir, "glean_data")
         handle = LibGleanFFI.INSTANCE.glean_initialize(this.gleanDataDir.path, applicationContext.packageName)
 
+        // If any pings were registered before initializing, do so now
+        this.pingTypeQueue.forEach { this.registerPingType(it) }
+        if (!Dispatchers.API.testingMode) {
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Why do we keep this queue around in testing mode?

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 11]
```diff
@@ -305,6 +304,30 @@ open class GleanInternalAPI internal constructor () {
...
+     * Send a list of pings.
+     *
+     * The ping content is assembled as soon as possible, but upload is not
+     * guaranteed to happen immediately, as that depends on the upload
+     * policies.
+     *
+     * If the ping currently contains no content, it will not be sent.
+     *
+     * @param pings List of pings to send.
+     * @return The async Job performing the work of assembling the ping
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Is `Job` here uppercased because that's the actual type of what `sendPingsGeneric` returns?

File: glean-core/src/lib.rs [hunk 12]
```diff
@@ -154,6 +157,41 @@ impl Glean {
...
+    }
+
+    pub fn register_ping_type(&mut self, ping: &PingType) {
+        if self.ping_registry.contains_key(&ping.name) {
+            log::error!("Duplicate ping named {}", ping.name)
+        }
+
+        // TODO: This just clones the PingType.  This means we have a copy of the PingType
+        // for the ping registry and another managed by handles on the language-specific
+        // side, but these objects are small and this seems like a simple solution...
```
Comment (badboy) [inline glean-core/src/lib.rs]: For now they are small. We can optimize this when it becomes a problem.

**Comment by badboy** [comment]
One thing I forgot: Can you run `make cbindgen` to also update the C header?

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 1 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Good point.  I think it probably does.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 3 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Oh, shoot.  It should be there, just committed without removing my playing around here.

File: glean-core/src/metrics/ping.rs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    pub name: String,
+    pub include_client_id: bool,
+}
+
+impl PingType {
+    pub fn new<A: Into<String> + Copy>(name: A, include_client_id: bool) -> Self {
```
[the remark this answers is not in the document]
Comment (mdboom) [reply glean-core/src/metrics/ping.rs]: My bad -- I think this is a leftover from earlier struggles to get this to work.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 10 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: It's, uh, complicated...

PingTypes are registered by virtue of their being created.  They are singletons.  Therefore, they only get registered once (regardless of when `Glean.registerPings()` is called -- it only works the first time).  This is because we don't want to use introspection and don't want to have an explicit data structure listing all the available pings.

So... in testing, we want to "rerun" the queue every time.  Outside of testing, this is just wasted memory.

But I'd be happy to just delete this clause altogether.  It's not a lot of wasted memory.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 11 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Yes.  Should probably make this `[Job]` (which will link in the generated docs).

File: glean-core/src/lib.rs (hunk 12 above)
Comment (mdboom) [reply glean-core/src/lib.rs]: If you agree this is ok, it's ok for me.  I felt this was a little "we couldn't figure out how to disable the borrow checker, so we gave up", and there is probably a way to use handles here.  But the duplication isn't a real problem, other than very small amounts of memory, since they are immutable "data classes" essentially.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 1 above)
Comment (Dexterp37) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: This needs to be `@Synchronized` as well.

File: glean-core/src/metrics/ping.rs [hunk 14]
```diff
@@ -0,0 +1,28 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::error::Result;
+use crate::Glean;
+
+#[derive(Debug, Clone)]
+/// Stores information about a ping.
```
Comment (Dexterp37) [inline glean-core/src/metrics/ping.rs]: Can we explain why this is needed, e.g. to make it possible to send pings with partial information, for example, no client id.

**Comment by mdboom** [comment]
> Let's merge if CI is green. How comes that we're not getting which test is failing when running Android tests?

I'm not sure, but it seemed to be an intermittent failure and I can no longer reproduce.  :cat: :fire:

**Comment by Dexterp37** [comment]
> 
> 
> > Let's merge if CI is green. How comes that we're not getting which test is failing when running Android tests?
> 
> I'm not sure, but it seemed to be an intermittent failure and I can no longer reproduce. 🐱 🔥

OH NOES, THEY ARE BACK (and gone again)!
